File: app/src/main/java/com/sharpdroid/registro/API/RESTfulAPIService.java
package com.sharpdroid.registro.API;

import android.support.annotation.NonNull;

import com.sharpdroid.registro.Interfaces.API.Absences;
import com.sharpdroid.registro.Interfaces.API.Communication;
import com.sharpdroid.registro.Interfaces.API.CommunicationDescription;
import com.sharpdroid.registro.Interfaces.API.FileTeacher;
import com.sharpdroid.registro.Interfaces.API.Lesson;
import com.sharpdroid.registro.Interfaces.API.LessonSubject;
import com.sharpdroid.registro.Interfaces.API.Login;
import com.sharpdroid.registro.Interfaces.API.MarkSubject;
import com.sharpdroid.registro.Interfaces.API.Note;
import com.sharpdroid.registro.Interfaces.API.Scrutiny;

import java.util.List;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface RESTfulAPIService {

    @POST("login")
    @FormUrlEncoded
    Observable<Login> postLogin(
            @NonNull @Field("login") String login,
            @NonNull @Field("password") String password,
            @NonNull @Field("key") String key);

    @GET("files")
    Observable<List<FileTeacher>> getFiles();

    @GET("file/{id}/{cksum}/download")
    Observable<ResponseBody> getDownload(
            @NonNull @Path("id") String id,
            @NonNull @Path("cksum") String cksum);

    @GET("absences")
    Observable<Absences> getAbsences();

    @GET("marks")
    Observable<List<MarkSubject>> getMarks();

    @GET("subjects")
    Observable<List<LessonSubject>> getSubjects();

    @GET("subject/{id}/lessons")
    Observable<List<Lesson>> getLessons(
            @Path("id") int id);

    @GET("notes")
    Observable<List<Note>> getNotes();

    @GET("communications")
    Observable<List<Communication>> getCommunications();

    @GET("communication/{id}/desc")
    Observable<CommunicationDescription> getcommunicationDesc(
            @Path("id") int id);

    @GET("communication/{id}/download")
    Observable<ResponseBody> getcommunicationDownload(
            @Path("id") int id);

    @GET("scrutinies")
    Observable<List<Scrutiny>> getScrutines();
}

File: app/src/main/java/com/sharpdroid/registroelettronico/API/V1/RESTfulAPIService.kt
package com.sharpdroid.registroelettronico.API.V1

import com.sharpdroid.registroelettronico.Interfaces.API.*
import io.reactivex.Observable
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.http.*

internal interface RESTfulAPIService {

    @POST("login")
    @FormUrlEncoded
    fun postLogin(
            @Field("login") login: String,
            @Field("password") password: String,
            @Field("key") key: String): Observable<Login>

    @get:GET("files")
    val files: Observable<List<FileTeacher>>

    @GET("file/{id}/{cksum}/download")
    fun getDownload(
            @Path("id") id: String,
            @Path("cksum") cksum: String): Observable<Response<ResponseBody>>

    @get:GET("absences")
    val absences: Observable<Absences>

    @get:GET("marks")
    val marks: Observable<List<MarkSubject>>

    @get:GET("subject")
    val subjects: Observable<List<LessonSubject>>

    @GET("subject/{id}/lessons")
    fun getLessons(
            @Path("id") id: Int,
            @Query("teacherCode") teacherCode: String): Observable<List<Lesson>>

    @get:GET("notes")
    val notes: Observable<List<Note>>

    @get:GET("communications")
    val communications: Observable<List<Communication>>

    @GET("communication/{id}/desc")
    fun getCommunicationDesc(
            @Path("id") id: Int): Observable<CommunicationDescription>

    @GET("communication/{id}/download")
    fun getCommunicationDownload(
            @Path("id") id: Int): Observable<Response<ResponseBody>>

    @get:GET("scrutinies")
    val scrutines: Observable<List<Scrutiny>>

    @GET("events")
    fun getEvents(
            @Query(value = "start") start: Long,
            @Query(value = "end") end: Long): Observable<List<Event>>
}
